File: graph/207. Course Schedule.py
from collections import defaultdict
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        preMap = defaultdict(list)
        for crs, pre in prerequisites:
            preMap[crs].append(pre)

        # visitSet
        visiting = set() # all courses along curr DFS path

        def dfs(course):
            """
            Return if course can be completed
            """
            if course in visiting: # detect loop
                return False
            if len(preMap[course]) == 0:
                return True # can be completed without any prereq

            visiting.add(course)
            for pre in preMap[course]:
                if not dfs(pre):
                    return False
            preMap[course] = [] # shortcut: return True if ever visited
            visiting.remove(course)
            return True

        for c in range(numCourses): # iterate through every course
            if not dfs(c):
                return False
        return True

File: graph/207. Course Schedule.go
func dfs(course int, graph map[int][]int, visiting map[int]bool) bool {
    if visiting[course] {
        return false
    }
    if len(graph[course]) == 0 {
        return true
    }

    visiting[course] = true
    for _, post := range graph[course] {
        if !dfs(post, graph, visiting) {
            return false
        }
    }
    graph[course] = []int{} // shortcut: return true if ever visited
    visiting[course] = false
    return true
}

func canFinish(numCourses int, prerequisites [][]int) bool {
    graph := make(map[int][]int)
    // Check if there's cycle
    visiting := make(map[int]bool)
    // build the graph
    for _, tuple := range prerequisites {
        graph[tuple[1]] = append(graph[tuple[1]], tuple[0])
        visiting[tuple[1]] = false
    }
    // fmt.Printf("%v\n", graph)

    for c := range numCourses {
        if !dfs(c, graph, visiting) {
            return false
        }
    }

    return true
}
